README.txt
-----------------------------------------------------------------------------------
Project: OpenGTS - Open GPS Tracking System
URL    : http://www.opengts.org
File   : README.txt
-----------------------------------------------------------------------------------

Notes:
1) Please refer to the included "OpenGTS_Config.pdf" file for all installation and 
   configuration information for the OpenGTS system.
2) When updating to a newer version of OpenGTS, it is highly recommended that the 
   following command be run to update the database with the latest table/column 
   changes:
     $  bin/dbAdmin.pl -tables=ca
   Or, run the following if alternate key fields also need to be updated:
     $  bin/dbAdmin.pl -tables=cak
3) On Windows 7, Internet Explorer 8, if you receive the error 
      Error [jsMapInit:[object error]
   then you may need to enable "Active Scripting" in the security tab.
4) A listing of the database tables and fields can be found in the file "SCHEMA.txt"
   in the OpenGTS installation diretory.
5) Documentation is now available online at the following link:
     http://www.opengts.org/documentation.html
6) This open-source version of OpenGTS includes support for the following GPS 
   tracking devices:
     - Aspicore GSM Tracker (Nokia, Samsung, Sony Ericsson phones).
       [http://www.aspicore.com]
     - Sanav GC-101, MT-101, and CT-24 Personal Tracker (HTTP-based protocol).
       Sanav GX-101 Vehicle Tracker (HTTP-based protocol).
       [http://www.sanav.com]
     - V-Sun 3338 Personal Tracker.
     - Android App "GPS2OpenGTS"
       [http://www.androidzoom.com/android_applications/communication/gps2opengts_nfsk_download.html]
     - iCare G3300 Personal Tracker.
     - Most TK102/TK103 devices.
     - Xexun TK102.
     - ZhongShan SIPGEAR Technology Co, Ltd.
     - TAIP (Trimble ASCII Interface Protocol).
     - TrackStick GPS data logger.
     - "GPSMapper" capable phones.
     - "NetGPS" capable devices.
     - HP hw6965 Windows/CE phone (OpenDMTP compliant)
     - Mologogo capable phones
     - Various Boost Mobile phones (OpenDMTP compliant)

Please let us know if you find any issues with this release.

-----------------------------------------------------------------------------------

Visit the following link for additional downloadable documentation and add-ons for 
OpenGTS:
  http://www.opengts.org/documentation.html
  http://www.geotelematic.com/documentation.html

-----------------------------------------------------------------------------------

This package may contain one or more of the following source modules which contain 
copyrights from their respective authors.  Please review these source modules for 
their copyright and license information:
   war/track/js/mapstraction/mapstraction.js
   war/track/js/mapstraction/mapstraction-geocode.js
   war/track/js/mapstraction/labeledmarker.js
   war/track/js/sorttable/sorttable.js

-----------------------------------------------------------------------------------

Contributors:
  See the following link for a list of contributors:
    http://www.opengts.org/info.html

-----------------------------------------------------------------------------------

Sample data:
Additional documentation for installing sample 'demo' data into the database can be
found in the "README.txt" file in the "sampleData" directory at
    sampleData/README.txt

-----------------------------------------------------------------------------------

Boost Mobile Motorola phones:
For information regarding support for various Boost Mobile Motorola phones, please
see the document at 
    MotoDMTP/MotoDMTP.txt

-----------------------------------------------------------------------------------

Sanav GC-101 GPS Tracking Device Support:
Additional documentation for installing and configuring the GC-101 server within
OpenGTS can be found in the "README.txt" file in the "gc101" source directory at
   src/org/opengts/war/gc101/README.txt

See the following link for manufacturer's product information:
   http://www.sanav.com

-----------------------------------------------------------------------------------

TAIP (Trimble ASCII Interface Protocol) Support:
Support for TAIP (Trimble ASCII Interface Protocol) is included in this release.  
This server uses the raw-socket mode device communication server based on the example
'template' server.  The "taip.jar" server jar file can be built with the command:
   > ant taip
And can be started in the same manner that other servers are started using the
"runserver.pl" command as follows:
   > $GTS_HOME/bin/runserver.pl -s taip

-----------------------------------------------------------------------------------

ICare G3300 Personal GPS Tracking Device Support:
Support for the ICare G3300 GPS tracking device is included in this release.  This
server uses the raw-socket mode device communication server based on the example
'template' server.  The "icare.jar" server jar file can be built with the command:
   > ant icare
And can be started in the same manner that other servers are started using the
"runserver.pl" command as follows:
   > $GTS_HOME/bin/runserver.pl -s icare

-----------------------------------------------------------------------------------

Updating existing Geozone table entries 
(only required when upgrading from versions previous to v1.9.3):

If you have existing Geozone table entries, they will need to be updated to include
Bounding-Box information used to optimize Geozone lookups.  This update process is
done automatically with the following steps:

1) Update the new Geozone bounding-box table fields:
     > bin/dbAdmin.pl -tables=ca

2) For each account having Geozone entries, run the following command (replace "<accountID>" 
   with the appropriate account-IDs for each account which has Geozone entries):
     > bin/admin.pl Geozone -account=<accountID> -list -update

The update process should now be complete.

-----------------------------------------------------------------------------------

Runtime config file property key references
(for *.conf files, such as 'default.conf', 'common.conf', 'webapp.conf, ...):
- The prefixing "#" is a comment character, indicating that the following text is for
  informational purposes only, and should not be parsed.  Comments must be placed on
  a separate line, and must not be specified at the end of a property specification.
- Previously defined property keys may be referenced by placing them in ${...} brackets,
  such as ${log.name}.  Environment variables may also be referenced in this manner, as in 
  ${GTS_HOME}, or ${CATALINA_HOME}.
- A default value may be specified for referenced property keys which have not been defined.   
  For instance ${OUTPUT_DIR=/tmp} would resolve to "/tmp", if the property OUTPUT_DIR is not 
  defined (as an environment variable, or otherwise), or will resolve to the value specified 
  on the OUTPUT_DIR property (or environment variable) if it is defined.
- Recursive propery key references may also be specified.  For instance, the specification
  ${THIS_DIR=${THAT_DIR=/tmp}} will first attempt to resolve the value for "THIS_DIR", if not
  found, then value of THAT_DIR will be returned.  If THAT_DIR is not defined, then finally
  the specification will resolve to "/tmp".
- Property reference specifications for which the property is not defined, and there is no
  specified default value will resolve to the literal "${var}" string.  That is, if "var"
  is not defined, then ${var} will resolve to the literal string "${var}".
- Property reference specifications are only resolved at the time a property value is 
  requested, not at the time the property key=value line is parsed (ie. lazy resolution).  
  For instance, in the property specification below, "var_a" will resolve to "test", even 
  though "var_b" is not defined until after "var_a":
    var_a=${var_b}
    var_b=test
  This means that if "var_b" should ever change, the resolved value of "var_a" will also
  change accordingly.  [ie. "var_a" will only be resolved at the time a call to an RTConfig
  property value retrieval method is called - such as RTConfig.getString("var_a")].
  Write-only (assignable) property keys (ie. %log, %include) are the only exception to this 
  rule.  Since their assigned behavior is executed at the time they are parsed ('%log' prints
  a log message, and '%include' includes a file), write-only property keys must be assigned
  values that can be fully resolved at the time their values are assigned.
- Property keys may be re-assigned.  Property keys take on their last assigned value.  For
  instance, the following will display "Hello", then display "World":
    test=Hello
    %log=${test}
    test=World
    %log=${test}
    
The following are reserved read-only constant property keys:
   %version        returns the current version, as in "2.3.4"
   %contextName    returns the context 'name' (name of the servlet context, or main class name)
   %contextPath    returns the context 'path' (path to the servlet context, or $GTS_HOME)
   %configURL      returns the URL for the loaded config file
   %hostName       returns the current host-name
   %hostIP         returns the current mail host IP address

The following are reserved write-only (assignable) constant property keys:
   %log            displays the specified value to the log output
   %include        includes the config at the value URL (URL must exist, error otherwise)
   %include?       includes the config at the value URL (URL may exist, otherwise ignored)
   (The included URL protocol must be one of "file" or "http".)
   
The following are available within a PrivateLabel session context:
   session.name    The PrivateLabel Domain name
   session.locale  The PrivateLabel Domain locale

-----------------------------------------------------------------------------------

All trade names listed above are trademarks of their respective companies.
OpenGTS and GTS Enterprise are not affiliated with any of the listed companies.

-----------------------------------------------------------------------------------

Contact Info:
Please feel free to contact us regarding questions on this package.

Thanks,
Martin D. Flynn
devstaff@opengts.org


CHANGELOG.txt
-----------------------------------------------------------------------------------
GTS Changelog
Copyright 2007-2012 GeoTelematic Solutions, Inc.
-----------------------------------------------------------------------------------

Version 2.4.1 [2012/04/20]
- NEW: Added additional defined Status Codes.
- NEW: Ability to delete old events by DeviceGroup (as well as by Device). Also
       checks the very last event when deleting old events.  If the last event is
       prior to the deleted time range, the last event will be retained.
- NEW: Added additional columns types to EventDataLayout.java and FieldLayout.java.
- NEW: Added ability to hide "Unique ID", "SIM#", on Device Admin page via ACLs.
- NEW: Added "userInfo.showAccessControlList" property to hide ACL list on User Admin.
- NEW: Added field "simID" to Device record.
- NEW: Added command-line option to delete old EventData records by device group.
- NEW: Added command-line support for displaying "events-per-second" in EventData.
- NEW: Made GISGraphy street address lookup optional.
- NEW: Added Mapquest OSM/Arial lay support to OpenLayers MapProvider support. (See
       "openLayers" MapProvider definition for details.) Make sure to comply with 
       their terms of service.
- NEW: Added "Event Count by Day" device detail report.
- DCS: gprmc - update to be able to handle several protocol formats simultaneously.
- DCS: gc101 - fixed NPE issue when invalid $GPRMC record is parsed.
- DCS: template - "parseInsertRecord_ASCII_02" now allows using AccountID as MobileID.
- DCS: tk10x - added digital input change simulated events.
- FIX: Fixed typo - changed property key "groupInfo.deviceLisSortBy" name to 
       "groupInfo.deviceListSortBy".

Version 2.4.0 [2012/02/05]
- NEW: Added additional defined Status Codes.
- NEW: Added initial support for sending SMS messages from the OpenGTS
       "RuleFactoryExample.java" module (SMS outbound gateway config required).
- NEW: Added Economy-Units "L/100km" (Litres per 100 kilometers).
- NEW: Added "lastIgnitionOffTime" column to Device table.
- NEW: CheckInstall now shows report limits, SMS gateway configuration.
- UPD: Increased Resource table column "resourceID" length to 80 characters.
- UPD: Added "fuelRate" to EventData report layout, and additional fields to the
       Field report layout.
- UPD: "tinygeocoder" ReverseGeocodeProvider can now specify a "host" property.
- FIX: Fixed ReportTable.java where single device reports were sometimes not written
       to the output stream.
- FIX: Changed ForgotPassword.java to use default user-id with blank "User" entry.
- DCS: gc101 - Sanav HTTP-mode DCS now supports battery level, number of satellites, 
       and HDOP (if data is provided in event packet).
- DCS: tk10x - Added initial support for alternate TK103 format. Added support for
       TK102 battery-voltage and altitude.

Version 2.3.9 [2011/12/06]
- NEW: Added additional StatusCodes.
- NEW: Added "job" information to Device record.
- NEW: Added runtime config property "GoogleKML.pushpinBaseURL" to the "config.conf" 
       file to allow overriding the location where Google provides various pushpins.  
- CHG: Changed "CheckInstall" check for MySQL "max_onnections".
- CHG: Added check for Java vendor "Oracle" to CheckInstall.
- CHG: Added additional packet format support to example 'template' DCS module.
- CHG: Added UTF-8 characterset to SendMail support.

Version 2.3.8 [2011/10/10]
- NEW: Added additional StatusCodes.
- NEW: Added "map.showPushpins" property to MapProvider section to allow disabling
       the display of pushpins on the map.
- NEW: Added support for daylight-savings-time zone names to "DateTime.java".
- UPD: Added GeozoneID support to "GPSEvent.java"
- DCS: taip - Added estimated odometer calculation, and simulated geozone support.
- DCS: sipgear - default listen port changed from 31270 to 31170.
- DCS: icare - default listen port changed from 31260 to 31160.
- FIX: Fix memory leak issue in "OpenLayers.js" (thanks to YanXu "stelzbock")

Version 2.3.7 [2011/08/22]
- NEW: Added additional StatusCodes.
- CHG: Modified "tk10x" DCS to attempt to determine actual GMT timezone (configure
       TK103 devices relative to GMT+0, rather than the local timezone).
- CHG: Added unicode character encoding/decoding to address fields on map. 
- CHG: Changed Pushpin attribute "iconOffset" to "iconHostspot" (in private.xml).
       Support for "iconOffset" will still remain for this release.
- NEW: Added additional log messages to Geonames when reverse-geocoding fails.
- NEW: Unicode characters in addresses are now encoded when uploading to map page.
- CHG: Modified "setDate(...,Y,M,D)" (in "DateTime.java") to retain current time.
- FIX: Fixed FileTools "copyStreams" pattern matching (thanks to Jan Wedel, YanXu).
- FIX: Fixed issues with JSON parsing.
- UPD: Spanish language translation update (thanks to Sebastian Mantilla Beniers).

Version 2.3.6 [2011/07/15]
- NEW: Added new "tk10x" device communication server (DCS) module to support most
       TK102 and TK103 devices (also supports "sipgear" device protocol).  The
       "tk10x" DCS also support simulated geozone arrive/depart detection, and
       event code mapping to GTS status-codes. ("sipgear" DCS remains in this 
       version, but may be removed in a future release).
- CHG: Added "analog#" fields to report EventDataLayout.java
- CHG: Updated JSON format in "events.war" to remove quotes from numeric values.
- FIX: Fixed "template" DCS to remove redundant/incorrect "Begin TCP comm" message.
- FIX: Fixed EventDataLayout.java "batteryVolts" to use "event.getBatteryVolts()". 

Version 2.3.5 [2011/07/01]
- NEW: Added support for a new "MobileLocationProvider" API for integrating cell
       tower mobile location services.
- NEW: Added support for OpenCellID.  See "private.xml" (or "config.conf") for 
       configuration settings in the "MobileLocationProvider" section.

src/org/opengts/Version.java
// ----------------------------------------------------------------------------
// Copyright 2007-2012, GeoTelematic Solutions, Inc.
// All rights reserved
// ----------------------------------------------------------------------------
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
// http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ----------------------------------------------------------------------------
// Change History:
//  2007/01/25 v1.2.0  [Initial release]
//  2007/02/08 v1.2.1,  2007/02/11 v1.3.0,  2007/02/16 v1.3.1,  2007/02/18 v1.3.2
//  2007/02/26 v1.3.3,  2007/02/28 v1.3.4,  2007/03/11 v1.3.5,  2007/03/16 v1.3.6,
//  2007/03/25 v1.3.7,  2007/03/30 v1.4.0,  2007/04/15 v1.4.1,  2007/04/17 v1.4.2,
//  2007/05/06 v1.4.3,  2007/05/20 v1.4.4,  2007/05/25 v1.4.5,  2007/06/03 v1.4.6,
//  2007/06/13 v1.4.7,  2007/06/14 v1.4.8,  2007/06/30 v1.5.0,  2007/07/14 v1.5.1,
//  2007/07/27 v1.5.2,  2007/08/09 v1.5.3,  2007/09/16 v1.6.0,  2007/11/28 v1.6.1,
//  2007/12/13 v1.6.2,  2008/01/10 v1.6.3,  2008/02/04 v1.6.4,  2008/02/07 v1.6.5,
//  2008/02/11 v1.6.6,  2008/02/17 v1.6.7,  2008/02/21 v1.6.8,  2008/02/27 v1.6.9,
//  2008/03/12 v1.6.10, 2008/03/28 v1.7.0,  2008/04/11 v1.7.1,  2008/05/14 v1.8.0,
//  2008/05/20 v1.8.1,  2008/05/22 v1.8.2,  2008/06/20 v1.8.3,  2008/07/08 v1.8.4,
//  2008/07/21 v1.9.0,  2008/07/27 v1.9.1,  2008/08/08 v1.9.2,  2008/08/15 v1.9.3,
//  2008/08/17 v1.9.4,  2008/08/20 v1.9.5,  2008/08/24 v1.9.6,  2008/09/01 v1.9.7,
//  2008/09/12 v1.9.8,  2008/09/19 v2.0.1,  2008/10/16 v2.0.2,  2008/12/01 v2.0.3,
//  2009/01/01 v2.0.4,  2009/02/01 v2.0.5,  2009/02/20 v2.0.6,  2009/04/02 v2.0.7,
//  2009/05/01 v2.0.8,  2009/05/24 v2.0.9,  2009/05/27 v2.1.0,  2009/06/01 v2.1.1,
//  2009/07/01 v2.1.2,  2009/08/09 v2.1.3,  2009/08/23 v2.1.4,  2009/08/25 v2.1.4.1,
//  2009/09/23 v2.1.5,  2009/10/05 v2.1.6,  2009/11/01 v2.1.7,  2009/11/10 v2.1.8,
//  2009/12/16 v2.1.9,  2010/01/29 v2.2.0,  2010/04/11 v2.2.1,  2010/04/25 v2.2.2,
//  2010/05/24 v2.2.3,  2010/06/17 v2.2.4,  2010/07/04 v2.2.5,  2010/07/18 v2.2.6,
//  2010/09/10 v2.2.7,  2010/10/25 v2.2.8,  2010/11/29 v2.2.9,  2011/01/28 v2.3.0,
//  2011/03/08 v2.3.1,  2011/04/01 v2.3.2,  2011/05/15 v2.3.3,  2011/06/16 v2.3.4,
//  2011/07/01 v2.3.5,  2011/07/15 v2.3.6,  2011/08/21 v2.3.7,  2011/10/10 v2.3.8,
//  2011/12/06 v2.3.9,  2012/02/03 v2.4.0,  
//  2012/04/20 v2.4.1,
// ----------------------------------------------------------------------------
package org.opengts;

import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;
import java.text.SimpleDateFormat;

import org.opengts.util.*;

/**
*** This class provides runtime version information to the OpenGTS modules.
**/

public class Version
{

    // ------------------------------------------------------------------------

    // "public" as of 2.3.4-B06
    public  static final String COPYRIGHT = "Copyright 2007-2012, GeoTelematic Solutions, Inc.";

    // ------------------------------------------------------------------------

    // This string is parsed via 'grep' & 'sed' scripts and thus 
    // ONLY the version value specified within the quotes should change.
    private static final String VERSION = "2.4.1";

    // This public constant should only be accessed externally by the 'GTSAdmin' application.
    public  static final String GTS_ENTERPRISE_PREFIX = "E";
    public  static final String COMPILED_VERSION = 
        ////  GTS_ENTERPRISE_PREFIX + // "Enterprise"
        VERSION;

    // last compile time
    private static final long COMPILE_TIMESTAMP = CompileTime.COMPILE_TIMESTAMP;

    // ------------------------------------------------------------------------

    // package release time (modified by command-line 'sed' script to insert actual epoch time)
    private static final long PACKAGE_TIMESTAMP = 1334969129L;

    // ------------------------------------------------------------------------
    // ------------------------------------------------------------------------

    /**
    *** Initializes the RTConfig constant 'version' property.
    **/
    public static void initVersionProperty()
    {
        RTProperties constantProps = RTConfig.getRuntimeConstantProperties();
        if (!constantProps.hasProperty(RTKey.VERSION)) {
            constantProps.setProperty(RTKey.VERSION, Version.getVersion());
            //System.out.println("Set Version = " + Version.getVersion());
        } else {
            // already initialized, no need to reinitialize
        }
    };

    // ------------------------------------------------------------------------
    // ------------------------------------------------------------------------

    /**
    *** Returns the copyright notice
    *** @return The copyright notice
    **/
    public static String getCopyright()
    {
        return COPYRIGHT;
    }

    /** 
    *** Returns the compiled version String
    *** @return The version String
    **/
    public static String getVersion()
    {
        return COMPILED_VERSION;
    }

    /** 
    *** Returns the compiled package release timestamp
    *** @return The package release timestamp
    **/
    public static long getPackageTimestamp()
    {
        return PACKAGE_TIMESTAMP;
    }

    /** 
    *** Returns the last compiled timestamp
    *** @return The last compiled timestamp
    **/
    public static long getCompileTimestamp()
    {
        return COMPILE_TIMESTAMP;
    }
    
    /**
    *** Returns the last compiled time date string
    *** @return The last compiled time date string
    **/
    public static String getCompileTime()
    {
        return (new DateTime(COMPILE_TIMESTAMP)).toString();
    }

    // ------------------------------------------------------------------------

    /**
    *** Returns a String representation of this Version
    *** @return A String representation of this Version
    **/
    public static String getInfo()
    {
        StringBuffer sb = new StringBuffer();
        sb.append(Version.COPYRIGHT).append("\n");
        sb.append("Version: ").append(Version.getVersion()).append("\n");
        SimpleDateFormat dateFmt = new SimpleDateFormat("yyyy/MM/dd HH:mm:ss zzz", Locale.US);
        if (PACKAGE_TIMESTAMP > 0L) {
            String dfmt = dateFmt.format(new Date(PACKAGE_TIMESTAMP * 1000L));
            sb.append("Package: [" + PACKAGE_TIMESTAMP + "] " + dfmt + "\n");
        }
        if (COMPILE_TIMESTAMP > 0L) {
            String dfmt = dateFmt.format(new Date(COMPILE_TIMESTAMP * 1000L));
            sb.append("Compile: [" + COMPILE_TIMESTAMP + "] " + dfmt + "\n");
        }
        return sb.toString();
    }

    // ------------------------------------------------------------------------

    /**
    *** Main entry point to display the current version
    *** @param argv The command-line args
    **/
    public static void main(String argv[])
    {
        if ((argv.length > 0) && argv[0].equals("-info")) {
            // Version: 2.0.5
            // Compile: [1211915119] 2008/05/27 19:36:38 GMT
            // Package: [1211915119] 2008/05/27 19:36:38 GMT
            System.out.println(Version.getInfo());
        } else
        if ((argv.length > 0) && argv[0].equals("-package")) {
            // "1211915119"
            System.out.println(Version.getPackageTimestamp());
        } else
        if ((argv.length > 0) && argv[0].equals("-compile")) {
            // "1211915119"
            System.out.println(Version.getCompileTimestamp());
        } else {
            // "1.8.3"
            System.out.println(Version.getVersion());
        }
    }

    // ------------------------------------------------------------------------

}
